logical/logical.go

```diff
@@ -2,6 +2,32 @@ package logical
 
 import log "github.com/mgutz/logxi/v1"
 
+// BackendType is the type of backend that is being implemented
+type BackendType uint32
+
+// The these are the types of backends that can be derived from
+// logical.Backend
+const (
+	TypeUnknown     BackendType = 0 // This is also the zero-value for BackendType
+	TypeLogical     BackendType = 1
+	TypeCredential  BackendType = 2
+	TypePassthrough BackendType = 3
```

Comment:
How would type assertion work for plugin-based backends? I wouldn't be able to assert on `logical.PassthroughBackend` since the plugin doesn't implement that type.

Moving from using special case type to flags makes sense, but I am not sure what does that mean from the backend implementation perspective.